prow/cmd/job-forker/forking.go

```diff
@@ -0,0 +1,264 @@
+		repo,
+		len(presubmits),
+		len(postsubmits),
+		len(periodics),
+		version,
+		output,
+	)
+}
+
+func removeDeprecatedConfigs(repo, baseRepoJobsDirPath string, versions []string) {
+
+	repoString := strings.ReplaceAll(repo, "/", "-")
+	log.Printf("repoString: %v\n", repoString)
+	forkedDir := path.Join(baseRepoJobsDirPath, ForkDir)
+	forkedFiles, err := ghi.GetFileNames(forkedDir, []string{}, false)
+	if err != nil && !errors.Is(err, os.ErrNotExist) {
+		log.Fatalf("Couldn't read from forked files Directory: %v\n", err)
+	}
+	log.Printf("forkedFiles: %v\n", forkedFiles)
+
+	for _, forkedFile := range forkedFiles {
+		if !strings.Contains(forkedFile, repoString) {
+			log.Printf("%v didn't match %v\n", forkedFile, repoString)
+			continue
+		}
+		log.Printf("%v matched %v\n", forkedFile, repoString)
+		// branched File belongs to repo
+		matches := false
+		for _, version := range versions {
+			fileVersion := strings.ReplaceAll(version, ".", "-")
+
+			if strings.Contains(forkedFile, repoString+"-"+fileVersion) {
+				// branched File has corresponding branch
+				matches = true
+				break
+			}
+
+		}
+
+		if !matches {
+			// File is deprecated and has no corresponding branch to it anymore
+			log.Printf("Removing %v, because it's config is deprecated\n", forkedFile)
+			if err = os.Remove(forkedFile); err != nil {
+				log.Printf("warn: couldn't remove file %v\n", err)
+			}
+		}
+	}
+}
+
+func readReposFromJobs(files []string) []string {
+	repos := []string{}
+	for _, file := range files {
+		j, err := config.ReadJobConfig(file)
+		if err != nil {
+			log.Fatalf("Couldn't read jobConfig: %v\n", err)
+		}
+		for repo := range j.PresubmitsStatic {
+			if !contains(repos, repo) {
+				repos = append(repos, repo)
+			}
+		}
+
+		for repo := range j.PostsubmitsStatic {
+			if !contains(repos, repo) {
+				repos = append(repos, repo)
+			}
+		}
+
+		for _, periodic := range j.Periodics {
+			for _, ref := range periodic.ExtraRefs {
+				repo := ref.OrgRepoString()
+
+				if !contains(repos, repo) {
+					repos = append(repos, repo)
+				}
+			}
+		}
+	}
+
+	return repos
+}
+
+func generateForkedConfigurations(baseRepo *ghi.Repository) error {
+	baseRepoJobsDirPath := path.Join(baseRepo.RepoClient.Directory(), o.configsPath)
+	fileNames, err := ghi.GetFileNames(baseRepoJobsDirPath, []string{ForkDir}, o.ghio.Recursive)
+	log.Printf("Files in baseRepoJobsDirPath: %v\n", fileNames)
+	if err != nil {
+		return err
+	}
+
+	repos := readReposFromJobs(fileNames)
+	for _, repoString := range repos {
+		rep, err := ghi.NewRepository(repoString)
+		if err != nil {
+			return err
+		}
+
+		releaseBranches, err := rep.GetMatchingBranches(BranchPrefix + `v\d+\.\d+`)
+		if err != nil {
+			return err
+		}
+
+		log.Printf("There are %v release branches for repo %v\n", len(releaseBranches), rep.FullRepoName)
+		versions := generateVersionsFromBranches(releaseBranches, BranchPrefix)
```

Comment:
Why do you separate the version from the branch here and put them together again in the `generatePresubmits`... functions?  Please split it later, where you really need the version.